File: src-tauri/src/webvtt.rs
use regex::Regex;
use std::fmt::Debug;
use serde::{Deserialize, Serialize};

#[derive(PartialEq, Serialize, Deserialize, Clone)]
pub struct Caption {
    pub(crate) start_time: String,
    end_time: String,
    pub(crate) text: String,
}

impl Caption {
    pub fn parse_start_time(&self) -> u32 {
        let time_parts: Vec<&str> = self.start_time.split(':').collect();
        let hour: u32 = time_parts[0].parse().unwrap();
        let minute: u32 = time_parts[1].parse().unwrap();
        let second: Vec<&str> = time_parts[2].split('.').collect();
        let sec: u32 = second[0].parse().unwrap();
        let millisec: u32 = second[1].parse().unwrap();
        millisec + sec * 1000 + minute * 60 * 1000 + hour * 60 * 60 * 1000
    }
}

impl Debug for Caption {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "start: {}, end: {}, text: {}", self.start_time, self.end_time, self.text)
    }
}

pub fn parse_webvtt(webvtt: &str) -> Vec<Caption> {
    let mut captions = Vec::new();

    // Matches "HH:MM:SS.sss --> HH:MM:SS.sss"
    let time_pattern = Regex::new(r"(\d{2}:\d{2}:\d{2}.\d{3}) --> (\d{2}:\d{2}:\d{2}.\d{3})").unwrap();

    let mut current_start_time = String::new();
    let mut current_end_time = String::new();
    let mut current_text = String::new();

    for line in webvtt.lines() {
        if time_pattern.is_match(line) {
            let caps = time_pattern.captures(line).unwrap();
            current_start_time = caps.get(1).unwrap().as_str().to_string();
            current_end_time = caps.get(2).unwrap().as_str().to_string();
        } else if (!line.is_empty()) && (line != "WEBVTT") {
            let caption = Caption {
                start_time: current_start_time.clone().to_string(),
                end_time: current_end_time.clone().to_string(),
                text: line.trim().to_string(),
            };
            captions.push(caption);
            current_text.clear();
        }
    }

    captions
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_webvtt() {
        let input = "WEBVTT

00:00:00.000 --> 00:00:09.300
Caption text

00:00:09.300 --> 00:00:18.140
More caption text";

        let result = parse_webvtt(input);

        assert_eq!(
            result, vec![
                Caption { start_time: "00:00:00.000".to_string(), end_time: "00:00:09.300".to_string(), text: "Caption text".to_string() },
                Caption { start_time: "00:00:09.300".to_string(), end_time: "00:00:18.140".to_string(), text: "More caption text".to_string() }
            ]
        );
    }
}

File: src/lib/webvtt.ts
export class Caption {
    constructor(public startTime: string, public endTime: string, public text: string) {
        this.startTime = startTime;
        this.endTime = endTime;
        this.text = text;
    }

    parseStartTimeMillis(): number {
        const timeParts: string[] = this.startTime.split(':');
        const hour: number = parseInt(timeParts[0]);
        const minute: number = parseInt(timeParts[1]);
        const secondParts: string[] = timeParts[2].split('.');
        const second: number = parseInt(secondParts[0]);
        const millisec: number = parseInt(secondParts[1]);
        return millisec + second * 1000 + minute * 60 * 1000 + hour * 60 * 60 * 1000;
    }
}

export function parseWebVTT(webvtt: string): Caption[] {
    const captions: Caption[] = [];
    const timePattern = /(\d{2}:\d{2}:\d{2}\.\d{3}) --> (\d{2}:\d{2}:\d{2}\.\d{3})/;

    let currentStartTime = '';
    let currentEndTime = '';
    let currentText = '';

    webvtt.split('\n').forEach((line) => {
        const match = line.match(timePattern);
        if (match) {
            currentStartTime = match[1];
            currentEndTime = match[2];
        } else if (line !== '' && line !== 'WEBVTT') {
            if (currentText !== '') {
                currentText += '\n';
            }
            currentText += line.trim();
        } else if (line === '' && currentText !== '') {
            captions.push(new Caption(currentStartTime, currentEndTime, currentText));
            currentText = '';
        }
    });

    // Handle case where the last caption doesn't end with a newline
    if (currentText !== '') {
        captions.push(new Caption(currentStartTime, currentEndTime, currentText));
    }

    return captions;
}

export function compactWebVtt(captions: Caption[]): Caption[] {
    const compactedCaptions: Caption[] = [];
    let previousCaption: Caption | null = null;

    for (const caption of captions) {
        if (previousCaption && caption.text === previousCaption.text) {
            // Update only the end time if the text is the same
            previousCaption.endTime = caption.endTime;
        } else {
            // Add a new caption if the text is different
            previousCaption = new Caption(caption.startTime, caption.endTime, caption.text);
            compactedCaptions.push(previousCaption);
        }
    }

    return compactedCaptions;
}
